### PR TITLE
Plot Area: Plot clip bug

### LINKED ISSUE #488: Plotly.relayout doesn't update the subplot clip paths
a pretty high priority bug.

http://codepen.io/etpinard/pen/grBRao

### PR DESCRIPTION
**Pull request by mdtusz**
Fixes #488. The d3 selection was only being created. This has added an update "join" on
the selection so the clips will be updated whenever relayout is called.

### CODE REVIEW

File: src/plot_api/plot_api.js [hunk 1]
```diff
@@ -2887,23 +2887,28 @@ function lsInner(gd) {
...
+
+        plotClip.enter().append('clipPath')
             .attr({
                 'class': 'plotclip',
                 'id': clipId
             })
-            .append('rect')
+            .append('rect');
+
+        plotClip.selectAll('rect')
```
Comment (etpinard) [inline src/plot_api/plot_api.js]: Nicely done. Man, _I gotta got better at noticing these as a reviewer_.

Comment (mdtusz) [inline src/plot_api/plot_api.js]: _I gotta get better at noticing these as a programmer_.

File: test/jasmine/tests/plot_api_test.js [hunk 2]
```diff
@@ -20,6 +20,36 @@ describe('Test plot api', function() {
...
+                        clipRect = plotClip.children[0],
+                        clipWidth = +clipRect.getAttribute('width'),
+                        clipHeight = +clipRect.getAttribute('height');
+
+                    expect(clipWidth).toBe(240);
+                    expect(clipHeight).toBe(220);
+                })
+                .then(done);
+        });
+    });
```
Comment (etpinard) [inline test/jasmine/tests/plot_api_test.js]: can you also add one test for `Plots.resize`, e.g. like in this http://codepen.io/etpinard/pen/vGVZxe ?

(that's how I noticed that bug after all).

**Comment by etpinard** [comment]
:dancer: after adding that one `Plots.resize` test.

File: src/plot_api/plot_api.js (hunk 1 above)
Comment (mdtusz) [inline src/plot_api/plot_api.js]: But really, I think it's just a symptom of the "d3 pattern". It's not explicitly clear and pretty tough to reason about _what exactly_ each selection/block is doing.

Comment (etpinard) [inline src/plot_api/plot_api.js]: Looks like d3 v4 will make this easier.

Taken from https://medium.com/@mbostock/what-makes-software-good-943557f8a488#.rb39u8abk

image

**Pull request merged**